loam_multisense

gtinchev opened the pull request:
Could we tag it as v1 and rename it to loam_multisense?

Videos available here:
https://github.com/ipab-slmc/wiki/issues/61#issuecomment-231030485

Still to test 90+ rotations.

Commit:
Ported loam_continuous to indigo.
Translated LaserScan to PointCloud2, as loam_continuous expects that as input.

Commit:
Added readme

Commit:
Fix README

gtinchev committed:
Update README.md

Commit:
Started commenting the loam_continuous code.
Algorithm now correctly determines whether there's a new sweep.
I.e. rotation speed is taken into account.
Video: https://www.dropbox.com/s/hfonr6ulzpskju9/loam_lidar_sweep_fixed.ogv?dl=0

Commit:
Registration now correctly identifies edges and flat areas. Added additional registration topics.

Commit:
Corner points stitched correctly, when robot is static.

Commit:
Added filters manually, and via yaml config. What filters were originally used?

Commit:
Cannot rqt_plot.

Commit:
Removed filters for now. Added plots for 2 more params - s and iterNum.

Commit:
Algorithm working without imu for translations. Rotations shift the map, accurate IMU should be able to compensate for that.

Commit:
Mapped corner points - not accounting for rotation change.
Mapped planar points - accounting correctly for rotation change.

Commit:
Added more different rviz configurations for debug purposes. Forward/backward motion with AccumulateIMUShift now works correctly. Rotations to be done.

Commit:
IMU coordinate system altered, still need to account for rotations.

Commit:
IMU implementation working to some extend

mauricefallon commented on src/scanRegistration.cpp:
```diff
@@ -832,9 +832,9 @@ void imuHandler(const sensor_msgs::Imu::ConstPtr& imuIn)
-    imuRoll[imuPointerLast] = roll;
-    imuPitch[imuPointerLast] = -pitch;
-    imuYaw[imuPointerLast] = -yaw;
+    imuRoll[imuPointerLast] = yaw;
```
@gtinchev : 
I don't think this isn't correct but I have a very good idea about what is going on here.
I think you are very close to resolving this. Lets talk in the afternoon

Commit:
The coordinate frame of husky's IMU translated (orientation + angular velocity).
IMU measurements are still to be incorporated(linear acceleration), as AccumulateIMUShift does nothing currently.

mauricefallon commented:
I'm going to leave this PR open until you've been able to get to the bottom of the IMU behaviour

I just had a look at the loam code and its rough! You have done well to make progress

mauricefallon commented on src/scanRegistration.cpp:
```diff
@@ -46,12 +50,26 @@ sensor_msgs::PointCloud2 laserCloudLast2;
-const int imuQueLength = 50;
+// GT changes 1/07/16 - IMU HZ
+// const int imuQueLength = 50;
+const int imuQueLength = 100;
```
keep this change in mind

mauricefallon commented on src/scanRegistration.cpp:
```diff
@@ -384,25 +492,33 @@ void laserCloudHandler(const sensor_msgs::PointCloud2ConstPtr& laserCloudIn2)
     float diff = diffX * diffX + diffY * diffY + diffZ * diffZ;
 
+    // incremental average calculation of the diff
+    m = m + ((diff-m)/(cloudSize-12));
```
added but not used?

gtinchev replied:
This was my addition to calculate incremental average in order to see what is a "normal" value for the diff. Need to clean it up once I'm done with the IMU behaviour (a simple rotation was not enough for the IMU, still debugging).

Commit:
Code cleanup. IMU integrated.

mauricefallon commented:
im going to merge this as it is partly working, but its impossible for me to analyse or suggest errors in a pull request this large.

its preferable to break them up into much small manageable units in future
